File: typescript/test-live.mjs
// Live smoke test for the TypeScript SDK against production.
// Read-heavy: issues exactly one demo token, verifies it, marks it used.
import { RightOS, RightOSError } from "./dist/index.js";

const demo = new RightOS({ apiKey: "rk_demo_00000000000000000000" });
const pub = new RightOS();

let failures = 0;
function check(name, ok, extra = "") {
  console.log(`${ok ? "OK " : "NG "} ${name}${extra ? " — " + extra : ""}`);
  if (!ok) failures++;
}

// public
const plans = await pub.listPlans();
check("listPlans returns 5 plans", plans.length === 5, `got ${plans.length}`);

const policy = await pub.getLocationPolicy("loc_ev");
check(
  "loc_ev policy is non-transferable",
  policy.policy.transferable === false
);

const defs = await pub.listPolicies();
check(
  "listPolicies returns 7 presets and >= 10 country overlays",
  Object.keys(defs.presets).length === 7 &&
    Object.keys(defs.countryOverlays).length >= 10,
  `presets=${Object.keys(defs.presets).length} overlays=${Object.keys(defs.countryOverlays).length}`
);

// auth error surfaces as RightOSError
let authErr = null;
try {
  await new RightOS({ apiKey: "rk_live_invalid" }).listLocations();
} catch (e) {
  authErr = e;
}
check(
  "invalid key throws RightOSError(401)",
  authErr instanceof RightOSError && authErr.status === 401,
  authErr?.code
);

// operator flow on demo org
const locations = await demo.listLocations();
check("demo listLocations > 0", locations.length > 0, `got ${locations.length}`);

const issued = await demo.issueToken({
  locationId: locations[0].id,
  title: "SDK live test (ts)",
});
check("issueToken returns walletUrl", issued.walletUrl.includes(issued.token.id));

const verified = await pub.verifyToken(issued.token.id, issued.verificationCode);
check("verifyToken -> success", verified.result === "success", verified.result);

const used = await demo.useToken(issued.token.id);
check("useToken -> used", used.status === "used", used.status);

const again = await pub.verifyToken(issued.token.id, issued.verificationCode);
check("re-verify -> already_used", again.result === "already_used", again.result);

// webhooks (v0.4.0): demo org can list but not register
const hooks = await demo.listWebhooks();
check("demo listWebhooks returns array", Array.isArray(hooks));
let whErr = null;
try {
  await demo.createWebhook({ url: "https://example.com/hook" });
} catch (e) {
  whErr = e;
}
check(
  "demo createWebhook throws 403 demo_org",
  whErr instanceof RightOSError && whErr.status === 403 && whErr.code === "demo_org",
  whErr?.code
);

// webhook signature helper (offline)
{
  const { createHmac } = await import("node:crypto");
  const secret = "whsec_selftest";
  const body = '{"id":"evt_1"}';
  const t = Math.floor(Date.now() / 1000);
  const v1 = createHmac("sha256", secret).update(`${t}.${body}`).digest("hex");
  check(
    "verifyWebhookSignature accepts valid signature",
    await RightOS.verifyWebhookSignature(secret, `t=${t},v1=${v1}`, body)
  );
  check(
    "verifyWebhookSignature rejects tampered body",
    !(await RightOS.verifyWebhookSignature(secret, `t=${t},v1=${v1}`, body + "x"))
  );
}

console.log(failures === 0 ? "ALL PASS" : `${failures} FAILURES`);
process.exit(failures === 0 ? 0 : 1);

File: python/test_live.py
"""Live smoke test for the Python SDK against production.

Read-heavy: issues exactly one demo token, verifies it, marks it used.
"""
import sys

from rightos import RightOS, RightOSError

demo = RightOS(api_key="rk_demo_00000000000000000000")
pub = RightOS()
failures = 0


def check(name, ok, extra=""):
    global failures
    print(f"{'OK ' if ok else 'NG '} {name}{' - ' + str(extra) if extra else ''}")
    if not ok:
        failures += 1


plans = pub.list_plans()
check("list_plans returns 5 plans", len(plans) == 5, len(plans))

policy = pub.get_location_policy("loc_ev")
check("loc_ev policy is non-transferable", policy["policy"]["transferable"] is False)

defs = pub.list_policies()
check(
    "list_policies returns 7 presets and >= 10 country overlays",
    len(defs["presets"]) == 7 and len(defs["countryOverlays"]) >= 10,
    f"presets={len(defs['presets'])} overlays={len(defs['countryOverlays'])}",
)

try:
    RightOS(api_key="rk_live_invalid").list_locations()
    check("invalid key raises RightOSError(401)", False)
except RightOSError as e:
    check("invalid key raises RightOSError(401)", e.status == 401, e.code)

locations = demo.list_locations()
check("demo list_locations > 0", len(locations) > 0, len(locations))

issued = demo.issue_token(location_id=locations[0]["id"], title="SDK live test (py)")
check("issue_token returns walletUrl", issued["token"]["id"] in issued["walletUrl"])

verified = pub.verify_token(issued["token"]["id"], issued["verificationCode"])
check("verify_token -> success", verified["result"] == "success", verified["result"])

used = demo.use_token(issued["token"]["id"])
check("use_token -> used", used["status"] == "used", used["status"])

again = pub.verify_token(issued["token"]["id"], issued["verificationCode"])
check("re-verify -> already_used", again["result"] == "already_used", again["result"])

# webhooks (v0.4.0): demo org can list but not register
hooks = demo.list_webhooks()
check("demo list_webhooks returns list", isinstance(hooks, list))
try:
    demo.create_webhook("https://example.com/hook")
    check("demo create_webhook raises 403 demo_org", False)
except RightOSError as e:
    check(
        "demo create_webhook raises 403 demo_org",
        e.status == 403 and e.code == "demo_org",
        e.code,
    )

# webhook signature helper (offline)
import hashlib
import hmac as _hmac
import json as _json
import time as _time

from rightos import verify_webhook_signature

_secret = "whsec_selftest"
_body = _json.dumps({"id": "evt_1"})
_t = int(_time.time())
_v1 = _hmac.new(_secret.encode(), f"{_t}.{_body}".encode(), hashlib.sha256).hexdigest()
check(
    "verify_webhook_signature accepts valid signature",
    verify_webhook_signature(_secret, f"t={_t},v1={_v1}", _body),
)
check(
    "verify_webhook_signature rejects tampered body",
    not verify_webhook_signature(_secret, f"t={_t},v1={_v1}", _body + "x"),
)

print("ALL PASS" if failures == 0 else f"{failures} FAILURES")
sys.exit(0 if failures == 0 else 1)
